## Bayesian Classifier

Problem statement:Predict the survival of passanger in Titanic ship using Naive Bayesian machine learning algorithm

In [2]:
import pandas as pd

In [12]:
#importing titanic dataset into pandas dataframe
df=pd.read_csv("titanic_input.csv")
df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

## Perform some data exploration

In [17]:
#Removing columns from input dataframe that does not have any impact on final prediction
df.drop(['PassengerId','Name','SibSp','Parch','Embarked'],axis=1,inplace=True)

In [20]:
df.drop(['Cabin'],axis=1,inplace=True)

In [22]:
df.drop(['Ticket'],axis=1,inplace=True)

In [23]:
df.head(5)

Survived  Pclass     Sex   Age     Fare
0         0       3    male  22.0   7.2500
1         1       1  female  38.0  71.2833
2         1       3  female  26.0   7.9250
3         1       1  female  35.0  53.1000
4         0       3    male  35.0   8.0500

## Checking total null values in each columns

In [27]:
df.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age         177
Fare          0
dtype: int64

From above statiscs we can say that 'Age' column has total 177 null values and other columns do not have any null value

## Data cleaning by imputing mode in place of null values in Age column

In [30]:
# find mode of column age
from statistics import mode
mode=mode(df.Age)

In [31]:
#impute mode value in place of null values
df["Age"].fillna(mode, inplace = True)

In [32]:
#Verify Age column for null values
df.isnull().sum()

Survived    0
Pclass      0
Sex         0
Age         0
Fare        0
dtype: int64

From above we can say that Age column has zero null values

In [33]:
df.head(5)

Survived  Pclass     Sex   Age     Fare
0         0       3    male  22.0   7.2500
1         1       1  female  38.0  71.2833
2         1       3  female  26.0   7.9250
3         1       1  female  35.0  53.1000
4         0       3    male  35.0   8.0500

## Data preprocessing

In [36]:
#Column Sex has categorial values we need to convert it into numerical values
df=pd.get_dummies(df)
df.head(5)

Survived  Pclass   Age     Fare  Sex_female  Sex_male
0         0       3  22.0   7.2500           0         1
1         1       1  38.0  71.2833           1         0
2         1       3  26.0   7.9250           1         0
3         1       1  35.0  53.1000           1         0
4         0       3  35.0   8.0500           0         1

Age column is split into two columns Sex_female,Sex_male

In [37]:
#drop Age column
df.drop(['Age'],axis=1,inplace=True)
df.head(5)

Survived  Pclass     Fare  Sex_female  Sex_male
0         0       3   7.2500           0         1
1         1       1  71.2833           1         0
2         1       3   7.9250           1         0
3         1       1  53.1000           1         0
4         0       3   8.0500           0         1

## Split data into training and testing dataset

In [38]:
from sklearn.model_selection import train_test_split

In [39]:
X_train, X_test, y_train, y_test = train_test_split(df[['Pclass','Fare','Sex_female','Sex_male']], df.Survived, test_size=0.33)


In [41]:
X_test.head(3)

Pclass      Fare  Sex_female  Sex_male
546       2   26.0000           1         0
311       1  262.3750           1         0
503       3    9.5875           1         0

## Creating model

In [43]:
from sklearn.naive_bayes import GaussianNB
model=GaussianNB()

## Training the model

In [44]:
model.fit(X_train,y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

## Prediction on test data

In [46]:
y_pred=model.predict(X_test)

## Evaluate the model

In [47]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

In [49]:
con_matrix=confusion_matrix(y_test,y_pred)
con_matrix

array([[154,  30],
       [ 29,  82]], dtype=int64)

In [50]:
accuracy=accuracy_score(y_test,y_pred)
accuracy

0.8

We can accuracy of the model is 80%

## Prediction for new event

In [53]:
new_event=model.predict([[2,26.0000,1,0]])
new_event

array([1], dtype=int64)

From avove we can say that when Pclass=2	Fare=26.000	 Sex_female=1	Sex_male=0 there are chances of survial of passanger